scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/Lambda.scala

```diff
@@ -0,0 +1,261 @@
+      bucket = input.metadataPackage.getHost
+      key = input.metadataPackage.getPath.drop(1)
+      metadataJson <- parseFileFromS3(s3Client, bucket, key)
+      _ <- log("Retrieving metadata.json from s3 bucket")
+
+      minimumAssetsAndFilesErrors <- MetadataJsonSchemaValidator.checkJsonForMinimumObjects(metadataJson)
+      _ <- log("Checking that JSON has at least one Asset and one File")
+
+      atLeastOneEntryWithSeriesAndNullParentErrors <- MetadataJsonSchemaValidator.checkJsonForAtLeastOneEntryWithSeriesAndNullParent(metadataJson)
+      _ <- log("Checking that JSON has at least one entry that has both a series and a null parent")
+
+      validatedEntries <- validateAgainstSchema(metadataJson) // all entries validated except for where the 'type' could not be found
+      _ <- log("Checking that each entry type in JSON matches a schema")
+
+      valueValidator = new MetadataJsonValueValidator
+      fileEntries = validatedEntries("File")
+      validatedLocations <- valueValidator.checkFileIsInCorrectS3Location(s3Client, fileEntries)
+
+      fileEntriesWithValidatedLocation = fileEntries.zip(validatedLocations).map { case (entries, locationNel) => entries + (location -> locationNel) }
+
+      (fileEntriesWithValidatedExtensions, metadataFileIds) = {
+        val idsOfMetadataFiles = validatedEntries("Asset").flatMap { entry =>
+          entry.collect {
+            case (fieldName, validatedValue) if fieldName == "originalMetadataFiles" =>
+              val listOfValues = validatedValue.getOrElse(Arr()).arr.toList
+              listOfValues.map(_.str)
+          }.flatten
+        }
+        val (metadataEntries, nonMetadataEntries) = fileEntriesWithValidatedLocation.partition { fileEntry =>
+          val fileEntryId = fileEntry(id).getOrElse(Str("idFieldHasErrorInIt")).str
+          idsOfMetadataFiles.contains(fileEntryId)
+        }
+        val metadataEntriesWithValidatedExtensions = valueValidator.checkFileNamesHaveExtensions(metadataEntries)
+
+        (metadataEntriesWithValidatedExtensions ::: nonMetadataEntries, idsOfMetadataFiles)
+      }
+      updatedEntries = validatedEntries ++ Map("File" -> fileEntriesWithValidatedExtensions)
+      allEntryIds = valueValidator.getIdsOfAllEntries(updatedEntries, metadataFileIds)
+      entriesWithValidatedUniqueIds = valueValidator.checkIfAllIdsAreUnique(updatedEntries, allEntryIds)
+      entriesWithValidatedUuids = valueValidator.checkIfAllIdsAreUuids(entriesWithValidatedUniqueIds)
+
+      entryTypesGrouped = allEntryIds.groupBy { case (_, entryType) => entryType }
+      entriesWithValidatedParentIds = valueValidator.checkIfEntriesHaveCorrectParentIds(entriesWithValidatedUuids, allEntryIds.toMap, entryTypesGrouped)
+      entriesCompletelyValidated = valueValidator.checkForCircularDependenciesInFolders(entriesWithValidatedParentIds)
+
+      jsonStructuralErrors = {
+        minimumAssetsAndFilesErrors.zip(LazyList.from(1)).map { (error, errorNum) =>
```

Comment:
Sure, good spot.

😄 I did it `from(1)` because I thought it would read better, "Error 1", "Error 2"  but since probably only programmers will be reading the errors, it's fine and shortens the code.